### ISSUE TITLE
linux server with 4 x 1080 ti ... getting this error . UnityEnvironmentException: The Unity environment took too long to respond.

### ISSUE THREAD

**Issue by artoul** [Unity-Technologies/ml-agents#71]
i have a server with centos and latest drivers ... and i am running the script to teach it ...
if i run the project but a build for macOS everything works fine and it learns ...
i don't have a screen to see what is happening on the server ...

i am 100% sure the brain is set to external ... and as i said if i run this on macOS it works

i am getting exactly this

------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
timeout                                   Traceback (most recent call last)
~/python/ml-agents-master/python/unityagents/environment.py in __init__(self, file_name, worker_id, base_port)
     84                 self._socket.listen(1)
---> 85                 self._conn, _ = self._socket.accept()
     86                 self._conn.setblocking(1)

/usr/lib64/python3.5/socket.py in accept(self)
    194         """
--> 195         fd, addr = self._accept()
    196         # If our type has the SOCK_NONBLOCK flag, we shouldn't pass it onto the

timeout: timed out

During handling of the above exception, another exception occurred:

UnityEnvironmentException                 Traceback (most recent call last)
<ipython-input-3-2f8f70d5848c> in <module>()
----> 1 env = UnityEnvironment(env_name)
      2 print(str(env))
      3 brain_name = env.brain_names[0]

~/python/ml-agents-master/python/unityagents/environment.py in __init__(self, file_name, worker_id, base_port)
     91                 "The Unity environment took too long to respond. Make sure {} does not need user interaction to launch "
     92                 "and that the Academy and the external Brain(s) are attached to objects in the Scene.".format(
---> 93                     str(file_name)))
     94         except UnityEnvironmentException:
     95             proc1.kill()

UnityEnvironmentException: The Unity environment took too long to respond. Make sure akram4 does not need user interaction to launch and that the Academy and the external Brain(s) are attached to objects in the Scene.

------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

**Comment by Maomao110** [comment]
how about restart the PPO files in the Jupyter?
I met this problem  after my first running successful. and when I run the script the second time , I got this error. So I close the PPO file in the Jupyter , and stop running the PPO file in the Running tag. After I do this , I reopen the PPO file , and Run All , it works well.

**Comment by awjuliani** [comment]
Hi @artoul,

Is your server running xorg? You will need a virtual display to render the environment too in order for it to train. See here for more details on setting that up: https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Training-on-Amazon-Web-Service.md

**Comment by bitBrainResearch** [comment]
I'd second MaoMao's reply.  Try resetting the PPO script.  If that doesn't work, try rebuilding.  I've had this error a lot and my best guess is it happens when I build my environment in unity but don't wait long enough before running the script. OR it's when I fail to restart the script.  I get intermittent errors for both of those things and can't remember which gives me what error.

**Comment by ririro93** [comment]
Hello I'm new to Unity and the ml-agent. I have the same error message and I'm clueless how to solve this. The same .exe file I wrote used to work exactly as intended untill yesterday. But for some reason I'm getting the error message below since this morning. And I'm sure that the brain type is set to external and have already tried both restarting the script and rebuilding the .exe file from unity. Can anyone help?

When I execute the .exe file I can see how nothing changes in the scene and the step count stays at zero. However, when I run this cell

env = UnityEnvironment(file_name=env_name, curriculum=curriculum_file)
print(str(env))
brain_name = env.external_brain_names[0]

in the .exe file that opens up, the step count stays zero but objects move around, which is supposed to happen when steps are proceeded.

---------------------------------------------------------------------------
timeout                                   Traceback (most recent call last)
D:\GitHub\ml-agents\python\unityagents\environment.py in __init__(self, file_name, worker_id, base_port, curriculum)
     98                 self._socket.listen(1)
---> 99                 self._conn, _ = self._socket.accept()
    100                 self._conn.settimeout(30)

C:\ProgramData\Anaconda3\envs\anujland\lib\socket.py in accept(self)
    204         """
--> 205         fd, addr = self._accept()
    206         # If our type has the SOCK_NONBLOCK flag, we shouldn't pass it onto the

timeout: timed out

During handling of the above exception, another exception occurred:

UnityTimeOutException                     Traceback (most recent call last)
<ipython-input-3-a1598afa4a92> in <module>()
----> 1 env = UnityEnvironment(file_name=env_name, curriculum=curriculum_file)
      2 print(str(env))
      3 brain_name = env.external_brain_names[0]

D:\GitHub\ml-agents\python\unityagents\environment.py in __init__(self, file_name, worker_id, base_port, curriculum)
    105                     "The Unity environment took too long to respond. Make sure {} does not need user interaction to "
    106                     "launch and that the Academy and the external Brain(s) are attached to objects in the Scene."
--> 107                     .format(str(file_name)))
    108 
    109             if "apiNumber" not in p:

UnityTimeOutException: The Unity environment took too long to respond. Make sure test1 does not need user interaction to launch and that the Academy and the external Brain(s) are attached to objects in the Scene.

**Comment by batu** [comment]
Make sure you build your game with the brain set to external.

Also, give it at least 3 to 4 tries before giving up. In my experience trying to run the environment multiple times can get that error to dissapear. Also, try including a randomized port selection, something like this:
```
port_save = random.randint(0,100)
env = UnityEnvironment(file_name = project_path, base_port = 6013 + port_save)
```
which will reduce the wait time in case of a misfire.

Good luck, enjoy life.

**Comment by ririro93** [comment]
@batu wow I tried your method and it didn't work the first few times. But randomly started working after a few more tries. Thank you so much! 👍

**Comment by mmattar** [comment]
Closing this out due to inactivity. We've also since released v0.3 of ML-Agents which contains several enhancements.

**Comment by ricklamers** [comment]
I also ran into this issue on Windows 10 - brain was set to external - only restarting Jupyter Lab solved the issue. 

The same Jupyter Lab session (although different Python kernel) was used to connect ML Agents to a different Unity environment (which did work) (maybe it caused the bug?).
